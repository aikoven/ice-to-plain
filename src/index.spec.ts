import {Ice} from 'ice';
import {isEqual} from 'lodash';

import {iceToPlain, iceToJson, iceFromPlain} from './index';
import {Test} from '../fixtures/Test';

function expectEqual(v1: any, v2: any) {
  // use jest equality to check fields and have nice error reporting
  expect(v1).toEqual(v2);
  // use lodash equality to check prototypes
  expect(isEqual(v1, v2)).toBe(true);
}

test('Long', () => {
  // long that can be converted to number
  const smallLong = new Ice.Long(2 ** 40);

  const smallLongPlain = iceToPlain(smallLong);
  expect(smallLongPlain).toMatchSnapshot();
  expect(iceToJson(smallLong)).toEqual(JSON.stringify(smallLongPlain));
  expectEqual(iceFromPlain(smallLongPlain), smallLong);

  // long that can't be converted to number
  const largeLong = new Ice.Long(20000000, 0);

  const largeLongPlain = iceToPlain(largeLong);
  expect(largeLongPlain).toMatchSnapshot();
  expect(iceToJson(largeLong)).toEqual(JSON.stringify(largeLongPlain));
  expectEqual(iceFromPlain(largeLongPlain), largeLong);
});

test('Enum', () => {
  const enumValue = Test.TheEnum.First;

  const enumValuePlain = iceToPlain(enumValue);
  expect(enumValuePlain).toMatchSnapshot();
  expect(iceToJson(enumValue)).toEqual(JSON.stringify(enumValuePlain));
  expectEqual(iceFromPlain(enumValuePlain), enumValue);
});

test('Map with string keys', () => {
  const map: Test.StringKeyDict = new Map();
  map.set(
    'lol',
    new Test.TheStruct(new Ice.Long(42), new Test.SmallClass(new Ice.Long(24))),
  );
  map.set(
    'kek',
    new Test.TheStruct(
      new Ice.Long(123),
      new Test.SmallClass(new Ice.Long(321)),
    ),
  );

  const mapPlain = iceToPlain(map);
  expect(mapPlain).toMatchSnapshot();
  expect(iceToJson(map)).toEqual(JSON.stringify(mapPlain));
  expectEqual(iceFromPlain(mapPlain), map);
});

test('Map with bool keys', () => {
  const map: Test.BoolKeyDict = new Map();
  map.set(
    true,
    new Test.TheStruct(new Ice.Long(42), new Test.SmallClass(new Ice.Long(24))),
  );
  map.set(
    false,
    new Test.TheStruct(
      new Ice.Long(123),
      new Test.SmallClass(new Ice.Long(321)),
    ),
  );

  const mapPlain = iceToPlain(map);
  expect(mapPlain).toMatchSnapshot();
  expect(iceToJson(map)).toEqual(JSON.stringify(mapPlain));
  expectEqual(iceFromPlain(mapPlain), map);
});

test('Map with int keys', () => {
  const map: Test.IntKeyDict = new Map();
  map.set(
    1,
    new Test.TheStruct(new Ice.Long(42), new Test.SmallClass(new Ice.Long(24))),
  );
  map.set(
    -1,
    new Test.TheStruct(
      new Ice.Long(123),
      new Test.SmallClass(new Ice.Long(321)),
    ),
  );

  const mapPlain = iceToPlain(map);
  expect(mapPlain).toMatchSnapshot();
  expect(iceToJson(map)).toEqual(JSON.stringify(mapPlain));
  expectEqual(iceFromPlain(mapPlain), map);
});

test('HashMap', () => {
  const hashMap: Test.StructKeyDict = new Ice.HashMap();

  hashMap.set(new Test.KeyStruct(42, 'foo', Test.TheEnum.First), 'lol');
  hashMap.set(new Test.KeyStruct(24, 'bar', Test.TheEnum.Second), 'kek');

  const hashMapPlain = iceToPlain(hashMap);
  expect(hashMapPlain).toMatchSnapshot();
  expect(iceToJson(hashMap)).toEqual(JSON.stringify(hashMapPlain));
  expect(hashMap.equals(iceFromPlain(hashMapPlain), isEqual)).toBe(true);
});

test('Set', () => {
  const set = new Set([1, 2, 3]);

  const setPlain = iceToPlain(set);
  expect(setPlain).toMatchSnapshot();
  expect(iceToJson(set)).toEqual(JSON.stringify(setPlain));
  expectEqual(iceFromPlain(setPlain), set);
});

test('Struct', () => {
  const struct = new Test.TheStruct(
    new Ice.Long(42),
    new Test.SmallClass(new Ice.Long(24)),
  );

  const structPlain = iceToPlain(struct);
  expect(structPlain).toMatchSnapshot();
  expect(iceToJson(struct)).toEqual(JSON.stringify(structPlain));
  expectEqual(iceFromPlain(structPlain), struct);
});

test('Exception', () => {
  const exception = new Test.TheError('kek');

  const exceptionPlain = iceToPlain(exception);
  expect(exceptionPlain).toMatchSnapshot();
  expect(iceToJson(exception)).toEqual(JSON.stringify(exceptionPlain));
  expectEqual(iceFromPlain(exceptionPlain), exception);
});

test('Value', () => {
  const value = new Test.SmallClass(new Ice.Long(42));

  const valuePlain = iceToPlain(value);
  expect(valuePlain).toMatchSnapshot();
  expect(iceToJson(value)).toEqual(JSON.stringify(valuePlain));
  expectEqual(iceFromPlain(valuePlain), value);
});

test('Proxy', () => {
  const communicator = Ice.initialize();
  const proxy = Ice.LocatorPrx.uncheckedCast(
    communicator.stringToProxy('Category/Name:tcp -h blabla -p 1234')!,
  );

  const proxyPlain = iceToPlain(proxy);
  expect(proxyPlain).toMatchSnapshot();
  expect(iceToJson(proxy)).toEqual(JSON.stringify(proxyPlain));
  expect(() => iceFromPlain(proxyPlain)).toThrowError();
});

test('recursive', () => {
  const obj = {
    string: '123',
    bool: true,
    int: 42,
    null: null,
    undefined: undefined,
    array: [Test.TheEnum.First, new Test.SmallClass(new Ice.Long(42))],
    nested: {
      error: new Test.TheError('kek'),
      struct: new Test.TheStruct(
        new Ice.Long(42),
        new Test.SmallClass(new Ice.Long(24)),
      ),
    },
    rich: new Test.RichClass(
      new Ice.Long(20000000, 0),
      new Test.TheStruct(
        new Ice.Long(42),
        new Test.SmallClass(new Ice.Long(24)),
      ),
      Test.TheEnum.Second,
      ['lol', 'kek'],
      [
        new Test.SmallClass(new Ice.Long(42)),
        new Test.SmallClass(new Ice.Long(24)),
      ],
      [
        new Test.TheStruct(
          new Ice.Long(42),
          new Test.SmallClass(new Ice.Long(24)),
        ),
      ],
      new Map([['foo', new Test.TheStruct()]]),
    ),
  };

  const objPlain = iceToPlain(obj);
  expect(objPlain).toMatchSnapshot();
  expect(iceToJson(obj)).toEqual(JSON.stringify(objPlain));
  expectEqual(iceFromPlain(objPlain), obj);
});

test('plain', () => {
  const obj = {
    undefined: undefined,
    function: () => {},
    number: 1,
    string: 'string',
    boolean: true,
    null: null,
    array: [undefined, 1, 'string', true, null, {lol: 'kek'}],
    nested: {lol: 'kek'},
  };

  expect(iceToJson(obj)).toEqual(JSON.stringify(obj));
});

test('keywords', () => {
  const obj = {
    await: new Test.Keywords._Await(1),
    break: new Test.Keywords._Break(),
    try: Test.Keywords.E._TRY,
    catch: new Test.Keywords._Catch(),
  };

  const objPlain = iceToPlain(obj);
  expect(objPlain).toMatchSnapshot();
  expect(iceToJson(obj)).toEqual(JSON.stringify(objPlain));
  expectEqual(iceFromPlain(objPlain), obj);
});
