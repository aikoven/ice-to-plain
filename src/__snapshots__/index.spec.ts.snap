// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Enum 1`] = `
Object {
  "@@type": "Test.TheEnum",
  "value": "First",
}
`;

exports[`Exception 1`] = `
Object {
  "@@type": "Test.TheError",
  "stringField": "kek",
}
`;

exports[`HashMap 1`] = `
Object {
  "@@type": "Ice.HashMap",
  "entries": Array [
    Object {
      "key": Object {
        "@@type": "Test.KeyStruct",
        "enumField": Object {
          "@@type": "Test.TheEnum",
          "value": "Second",
        },
        "intField": 24,
        "stringField": "bar",
      },
      "value": "kek",
    },
    Object {
      "key": Object {
        "@@type": "Test.KeyStruct",
        "enumField": Object {
          "@@type": "Test.TheEnum",
          "value": "First",
        },
        "intField": 42,
        "stringField": "foo",
      },
      "value": "lol",
    },
  ],
}
`;

exports[`Long 1`] = `
Object {
  "@@type": "Ice.Long",
  "value": 1099511627776,
}
`;

exports[`Long 2`] = `
Object {
  "@@type": "Ice.Long",
  "high": 20000000,
  "low": 0,
}
`;

exports[`Map with bool keys 1`] = `
Object {
  "@@type": "Map",
  "entries": Object {
    "false": Object {
      "@@type": "Test.TheStruct",
      "classField": Object {
        "@@type": "Test.SmallClass",
        "longField": Object {
          "@@type": "Ice.Long",
          "value": 321,
        },
      },
      "longField": Object {
        "@@type": "Ice.Long",
        "value": 123,
      },
    },
    "true": Object {
      "@@type": "Test.TheStruct",
      "classField": Object {
        "@@type": "Test.SmallClass",
        "longField": Object {
          "@@type": "Ice.Long",
          "value": 24,
        },
      },
      "longField": Object {
        "@@type": "Ice.Long",
        "value": 42,
      },
    },
  },
}
`;

exports[`Map with enum keys 1`] = `
Object {
  "@@type": "Map",
  "entries": Object {
    "{\\"@@type\\":\\"Test.TheEnum\\",\\"value\\":\\"First\\"}": "second",
  },
}
`;

exports[`Map with int keys 1`] = `
Object {
  "@@type": "Map",
  "entries": Object {
    "-1": Object {
      "@@type": "Test.TheStruct",
      "classField": Object {
        "@@type": "Test.SmallClass",
        "longField": Object {
          "@@type": "Ice.Long",
          "value": 321,
        },
      },
      "longField": Object {
        "@@type": "Ice.Long",
        "value": 123,
      },
    },
    "1": Object {
      "@@type": "Test.TheStruct",
      "classField": Object {
        "@@type": "Test.SmallClass",
        "longField": Object {
          "@@type": "Ice.Long",
          "value": 24,
        },
      },
      "longField": Object {
        "@@type": "Ice.Long",
        "value": 42,
      },
    },
  },
}
`;

exports[`Map with string keys 1`] = `
Object {
  "@@type": "Map",
  "entries": Object {
    "\\"kek\\"": Object {
      "@@type": "Test.TheStruct",
      "classField": Object {
        "@@type": "Test.SmallClass",
        "longField": Object {
          "@@type": "Ice.Long",
          "value": 321,
        },
      },
      "longField": Object {
        "@@type": "Ice.Long",
        "value": 123,
      },
    },
    "\\"lol\\"": Object {
      "@@type": "Test.TheStruct",
      "classField": Object {
        "@@type": "Test.SmallClass",
        "longField": Object {
          "@@type": "Ice.Long",
          "value": 24,
        },
      },
      "longField": Object {
        "@@type": "Ice.Long",
        "value": 42,
      },
    },
  },
}
`;

exports[`Proxy 1`] = `
Object {
  "@@type": "Ice.LocatorPrx",
  "value": "Category/Name -t -e 1.1:tcp -h blabla -p 1234 -t 60000",
}
`;

exports[`Set 1`] = `
Object {
  "@@type": "Set",
  "value": Array [
    1,
    2,
    3,
  ],
}
`;

exports[`Struct 1`] = `
Object {
  "@@type": "Test.TheStruct",
  "classField": Object {
    "@@type": "Test.SmallClass",
    "longField": Object {
      "@@type": "Ice.Long",
      "value": 24,
    },
  },
  "longField": Object {
    "@@type": "Ice.Long",
    "value": 42,
  },
}
`;

exports[`Value 1`] = `
Object {
  "@@type": "Test.SmallClass",
  "longField": Object {
    "@@type": "Ice.Long",
    "value": 42,
  },
}
`;

exports[`keywords 1`] = `
Object {
  "await": Object {
    "@@type": "Test.Keywords.Await",
    "import": 1,
  },
  "break": Object {
    "@@type": "Test.Keywords.Break",
    "yield": "yield",
  },
  "catch": Object {
    "@@type": "Map",
    "entries": Object {},
  },
  "try": Object {
    "@@type": "Test.Keywords.Continue",
    "value": "TRY",
  },
}
`;

exports[`recursive 1`] = `
Object {
  "array": Array [
    Object {
      "@@type": "Test.TheEnum",
      "value": "First",
    },
    Object {
      "@@type": "Test.SmallClass",
      "longField": Object {
        "@@type": "Ice.Long",
        "value": 42,
      },
    },
  ],
  "bool": true,
  "int": 42,
  "nested": Object {
    "error": Object {
      "@@type": "Test.TheError",
      "stringField": "kek",
    },
    "struct": Object {
      "@@type": "Test.TheStruct",
      "classField": Object {
        "@@type": "Test.SmallClass",
        "longField": Object {
          "@@type": "Ice.Long",
          "value": 24,
        },
      },
      "longField": Object {
        "@@type": "Ice.Long",
        "value": 42,
      },
    },
  },
  "null": null,
  "rich": Object {
    "@@type": "Test.RichClass",
    "enumField": Object {
      "@@type": "Test.TheEnum",
      "value": "Second",
    },
    "longField": Object {
      "@@type": "Ice.Long",
      "high": 20000000,
      "low": 0,
    },
    "optionalField": undefined,
    "primitiveKeyDictField": Object {
      "@@type": "Map",
      "entries": Object {
        "\\"foo\\"": Object {
          "@@type": "Test.TheStruct",
          "classField": null,
          "longField": Object {
            "@@type": "Ice.Long",
            "value": 0,
          },
        },
      },
    },
    "primitiveSeqField": Array [
      "lol",
      "kek",
    ],
    "smallClassSeqField": Array [
      Object {
        "@@type": "Test.SmallClass",
        "longField": Object {
          "@@type": "Ice.Long",
          "value": 42,
        },
      },
      Object {
        "@@type": "Test.SmallClass",
        "longField": Object {
          "@@type": "Ice.Long",
          "value": 24,
        },
      },
    ],
    "structField": Object {
      "@@type": "Test.TheStruct",
      "classField": Object {
        "@@type": "Test.SmallClass",
        "longField": Object {
          "@@type": "Ice.Long",
          "value": 24,
        },
      },
      "longField": Object {
        "@@type": "Ice.Long",
        "value": 42,
      },
    },
    "structSeqField": Array [
      Object {
        "@@type": "Test.TheStruct",
        "classField": Object {
          "@@type": "Test.SmallClass",
          "longField": Object {
            "@@type": "Ice.Long",
            "value": 24,
          },
        },
        "longField": Object {
          "@@type": "Ice.Long",
          "value": 42,
        },
      },
    ],
  },
  "string": "123",
  "undefined": undefined,
}
`;
